src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,401 @@
+        let! t, (version, events) = mergeBatches batches |> Stopwatch.Time
+        log |> logBatchRead streamName t events batchSize version
+        return version, events }
+
+module Token =
+    let create streamVersion : StreamToken =
+        { value = box streamVersion
+          version = streamVersion
+          streamBytes = -1 }
+    let supersedes struct (current, x) =
+            x.version > current.version
+type MessageDbConnection(readConnection, [<O; D(null)>]?writeConnection, [<O; D(null)>]?readRetryPolicy, [<O; D(null)>]?writeRetryPolicy) =
+    member _.ReadConnection = readConnection
+    member _.ReadRetryPolicy = readRetryPolicy
+    member _.WriteConnection = defaultArg writeConnection readConnection
+    member _.WriteRetryPolicy = writeRetryPolicy
+
+type BatchOptions(getBatchSize : Func<int>, [<O; D(null)>]?batchCountLimit) =
+    new (batchSize) = BatchOptions(fun () -> batchSize)
+    member _.BatchSize = getBatchSize.Invoke()
+    member _.MaxBatches = batchCountLimit
+
+[<RequireQualifiedAccess; NoComparison; NoEquality>]
+type GatewaySyncResult = Written of StreamToken | ConflictUnknown
+
+type MessageDbContext(connection : MessageDbConnection, batchOptions : BatchOptions) =
+    let isResolvedEventEventType (tryDecode, predicate) (e : ResolvedEvent) = predicate (tryDecode e.Data)
+    let tryIsResolvedEventEventType predicateOption = predicateOption |> Option.map isResolvedEventEventType
+    new (   connection : MessageDbConnection,
+            // Max number of Events to retrieve in a single batch. Also affects frequency of RollingSnapshots. Default: 500.
+            [<O; D null>] ?batchSize) =
+        MessageDbContext(connection, BatchOptions(batchSize = defaultArg batchSize 500))
+    member val BatchOptions = batchOptions
+
+    member _.TokenEmpty = Token.create -1L
+    member _.LoadBatched streamName log (tryDecode, isCompactionEventType) : Async<StreamToken * 'event[]> = async {
+        let! version, events = Read.loadForwardsFrom log connection.ReadRetryPolicy connection.ReadConnection batchOptions.BatchSize batchOptions.MaxBatches streamName 0L
+        return Token.create version, Array.chooseV tryDecode events }
+    member _.LoadFromToken useWriteConn streamName log token tryDecode
+        : Async<StreamToken * 'event[]> = async {
+        let streamPosition = token.version + 1L
+        let connToUse = if useWriteConn then connection.WriteConnection else connection.ReadConnection
+        let! version, events = Read.loadForwardsFrom log connection.ReadRetryPolicy connToUse batchOptions.BatchSize batchOptions.MaxBatches streamName streamPosition
+        return Token.create (max token.version version), Array.chooseV tryDecode events
+      }
+    member _.TrySync log streamName (token) (events, encodedEvents : EventData array) : Async<GatewaySyncResult> = async {
```

Comment:
I've standardised on using `x array` even if that's actually not per the F# style guide (I thought it was at one point but looking at it now). For better or worse that should thus also be used in this and in mdbclient.fs